src/main/groovy/org/asciidoctor/groovydsl/AsciidoctorExtensions.groovy

```diff
@@ -15,94 +15,177 @@
+    void registerExtensionsWith(Asciidoctor asciidoctor) {
         AsciidoctorExtensionHandler extensionHandler = new AsciidoctorExtensionHandler(asciidoctor)
-        for (def it : REGISTERED_EXTENSIONS) {
-            DelegatingScript script
+        for (def it : registeredExtensions) {
             switch (it) {
                 case Closure:
                     try {
-                        it.delegate = extensionHandler
-                        it.call()
+                        ((Closure) it).delegate = extensionHandler
+                        ((Closure) it).call()
                     } catch (e) {
-                        REGISTERED_EXTENSIONS.clear()
                         throw new AsciidoctorExtensionException("Error registering extension from class in ${it.class.name}", e)
                     }
                     break
                 case String:
-                    def shell = makeGroovyShell()
-                    script = shell.parse(it)
+                    GroovyShell shell = makeGroovyShell()
+                    DelegatingScript script = (DelegatingScript) shell.parse((String) it)
                     script.setDelegate(extensionHandler)
                     try {
                         script.run()
                     } catch (e) {
-                        REGISTERED_EXTENSIONS.clear()
+                        registeredExtensions.clear()
                         throw new AsciidoctorExtensionException('Error registering extension from string', e)
                     }
                     break
                 case File:
-                    def shell = makeGroovyShell()
-                    FileReader reader = new FileReader(it)
-                    script = shell.parse(reader, it.name)
-                    script.setDelegate(extensionHandler)
-                    try {
-                        script.run()
-                    } catch (e) {
-                        REGISTERED_EXTENSIONS.clear()
-                        throw new AsciidoctorExtensionException("Error registering extension from file ${it.name}", e)
-                    } finally {
-                        reader.close()
+                    File file = (File) it
+                    GroovyShell shell = makeGroovyShell()
+                    file.withReader { reader ->
+                        DelegatingScript script = (DelegatingScript) shell.parse(reader, file.name)
+                        script.setDelegate(extensionHandler)
+                        try {
+                            script.run()
+                        } catch (e) {
+                            throw new AsciidoctorExtensionException("Error registering extension from file ${file.name}", e)
+                        }
+                    }
+                    break
+                case Path:
+                    Path path = (Path) it
+                    GroovyShell shell = makeGroovyShell()
+                    Files.newBufferedReader(path).withReader { reader ->
```

Comment:
@ysb33r I just tried to build on Java 7 and it failed because the method `Files.newBufferedReader(Path)` only exists since Java 8.
Should we fix it so that it works with Java 7?
Otherwise I would be fine with bumping the requirements to Java 8 as well.
Building with Java 7 on the CI is more and more becoming a pain anyway.